lib/resty/upstream/healthcheck.lua

```diff
@@ -92,6 +93,92 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
     end
 end
 
+local function get_unchecked_peers(peers)
+    local peers_to_check = {}
+    for _, peer in ipairs(peers) do
+        if not checked_peers_upstream[peer.name] then
+            table.insert(peers_to_check, peer)
+        end
+    end
+    return peers_to_check
+end
+
+local function get_checked_peers(peers)
+    local checked_peers = {}
+    for _, peer in ipairs(peers) do
+        if checked_peers_upstream[peer.name] then
+            table.insert(checked_peers, peer)
+        end
+    end
+    return checked_peers
+end
+
+local function get_all_peers(u)
+    local primary_peers, err = upstream.get_primary_peers(u)
+    if not primary_peers then
+        errlog("failed to get primary peers: ", err)
+        return
+    end
+
+    local backup_peers, err = upstream.get_backup_peers(u)
+    if not backup_peers then
+        errlog("failed to get backup peers: ", err)
+        return
+    end
+
+    return primary_peers, backup_peers
+end
+
+local function get_all_unchecked_peers(u)
+    local primary_peers, backup_peers = get_all_peers(u)
+
+    local ppeers_to_check = get_unchecked_peers(primary_peers)
+    local bpeers_to_check = get_unchecked_peers(backup_peers)
+
+    return ppeers_to_check, bpeers_to_check
+end
+
+local function get_all_checked_peers(u)
+    local primary_peers, backup_peers = get_all_peers(u)
+
+    local ppeers_checked = get_checked_peers(primary_peers)
+    local bpeers_checked = get_checked_peers(backup_peers)
+
+    return ppeers_checked, bpeers_checked
+end
+
+local function set_status(peers, is_backup, ctx)
+    local dict = ctx.dict
+    local u = ctx.upstream
+
+    for i = 1, #peers do
+        local peer = peers[i]
+        local already_set = dict:get(gen_peer_key("ok:", u, is_backup, peer.id))
+        if already_set then
+            return
```

Comment:
The early return here is incorrect (afaict) - it skips processing the remaining peers as soon as it finds one that is set. I think you want a `continue` here, but Lua doesn't provide that.